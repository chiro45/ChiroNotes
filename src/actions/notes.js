import Swal from "sweetalert2";


import { db } from "../firebase/firebaseConfing";

import { types } from "../types/types";

import {loadNotes} from '../helpers/loadNotes'
import { fileUpload } from "../helpers/fileUpload";




export const startNewNote = ()=>{
    return async (dispatch, getState)=>{       //el getState nos trae todo el state de la aplicacion
        const {uid} = getState().auth;   
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        const doc =  await db.collection(`${uid}/journal/notes`).add(newNote)     
        
        dispatch(activeNote(doc.id, newNote))

        dispatch(addNewNote(doc.id, newNote))
    }
}

export const activeNote = (id,note )=>({
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
});

export const addNewNote = (id, note)=>({
        type: types.notesAddNew,
        payload:{
            id,
            ...note
        }
})


export const startLoadingNotes = (uid) =>{
    return async (dispatch)=>{
        const notes = await loadNotes(uid)
      
        dispatch(setNotes(notes))
    }
}
export const setNotes = (notes)=>({
    type: types.notesLoad,
    payload: notes
})


 export const startSaveNote = (note)=>{
    return async(dispatch, getState)=>{
        try {
            const {uid} = getState().auth;
        if(!note.url){
            delete note.url;
        }
        const noteToFirestore ={...note}

        delete noteToFirestore.id;

        if(note.title.length > 1){
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch(refreshNote(note.id,note ))

        Swal.fire('Guardado!', note.title, "success")
        }else{
            Swal.fire('Deberia tener un titulo!', note.title, "error")
        }

        } catch (error) {
            Swal.fire('Error en el guardado!',"hubo un problema, intenta nuevamente", "error")
        }


    }
}

export const refreshNote = (id, note) =>({
    type: types.notesUpdated,
    payload:{
        id,
        note:{
                id,
                ...note
        }

        
    }
})


export const startUploading =(file) =>{
    return async(dispatch, getState)=>{
        const {active:activeNote} = getState().notes

        Swal.fire({
            title: 'Subiendo...',
            text: 'Por Favor Espere...',
            allowOutsideClick: false,
            didOpen:()=>{
                Swal.showLoading()
            }
            
        });
       
        const fileUrl = await fileUpload(file)
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}


export const startDeleting = (id)=>{
    return async(dispatch, getState)=>{
        
        const uid = getState().auth.uid;

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id))

        Swal.fire('Eliminado!','', "success")
    }
}


export const deleteNote = (id)=>({
    type: types.notesDelete,
    payload: id
})


export const noteLogout = () =>({
    type: types.notesLogoutCleaning
   
})
