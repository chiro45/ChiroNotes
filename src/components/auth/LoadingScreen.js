import React from 'react';

export const LoadingScreen = () => {
  return (
    <div className='auth__main'>
    <div className='auth__box-container'>
        <h2 className='auth__loading-title'>Cargando <i className=" spinner fas fa-spinner"></i></h2>
    </div>
    </div>
  );
};
