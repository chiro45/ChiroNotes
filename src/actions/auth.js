import Swal from 'sweetalert2'


import { types } from "../types/types"
import {firebase, googleAuthProvider} from '../firebase/firebaseConfing'
import { finishLoading, startLoading } from "./ui"
import { noteLogout} from './notes'
//acciones que son usadas en el auth




export const startLoginEmailPassword = (email,password)=>{
    return(dispatch)=>{
        
        dispatch(startLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
        .then( ({ user }) =>{


            dispatch(login(user.uid, user.displayName))
            
            
        dispatch(finishLoading())

        })
        .catch(e => {
        Swal.fire("Error","El usuario o contraseÃ±a son incorrectos","error")
    })
}}
export const startRegisterWhithEmailPasswordName = (email,password, name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
        .then( async({ user })=>{

            await user.updateProfile({
                displayName: name

            })

            dispatch(
                login(user.uid, user.displayName)
             )
            
         
        })
        .catch(e => {
            Swal.fire("Error","El usuario ya ha sido registrado","error")
        
    })

}}


export const startGoogleLogin = ()=>{
    return (dispatch)=>{
       firebase.auth().signInWithPopup(googleAuthProvider) 
       .then(({ user })=>{

           dispatch(
               login(user.uid, user.displayName, user.photoURL),

               console.log(user.photoURL)
                 
               )
            })
            
       
       
       
    }
}


export const login = (uid, displayName, photoURL)=>{
        return{
            type: types.login,
            payload: {
                uid,
                photoURL,
                displayName,
            }
        }
}


export const startLogout = () =>{
    return async(dispatch)=>{
        firebase.auth().signOut()

        dispatch(logout())
        
        dispatch(noteLogout())
    }
}

export const logout = () =>({
    type: types.logout
})


