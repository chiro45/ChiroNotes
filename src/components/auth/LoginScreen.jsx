import  { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm';
import validator  from 'validator';
import { removeError, setError } from '../../actions/ui';



export const LoginScreen = () => {
                   
    const [show, setShow] = useState(false);                            //uso el state para mostrar la contraseña o no
    const {msgError} = useSelector(state=>state.ui)                     //usamos el hook use selector para extraer el msgError del store
    const dispatch = useDispatch();                                     //extraemos el useDispatch


    const [formValues, handleInputChange] = useForm({                   //coustomHook para el manejo del formulario
        email: '',
        password:''
    });

    const {email, password} = formValues;                               //desestructuracion de los datos
    
    const isFormValid = ()=>{                                           //validacion del formulario
        if(!validator.isEmail(email)){
            dispatch(setError('El email es incorrecto'))
            return false;
        }else if( password.length < 8){
            dispatch(setError('La contraseña es menor a 8 caracteres o estas no coinciden'))
            return false;
        }
        dispatch(removeError())
        
        return true;
    }

    const handleLogin = (e)=>{                                           //realiza el envio del formulario, si tiene errores los muestra y sino realiza el dispatch
       e.preventDefault();
       if(isFormValid()){
        dispatch(startLoginEmailPassword(email, password))

       }
    }

    const handleGoogleLogin = ()=>{                                      //en el caso del login con google nos permite realizarla
        dispatch(startGoogleLogin())
    }
    const showOrHiddenPassword = ()=> setShow(!show);                    //funcion para mostrar o no la contraseña

    return (
       
        
        <div>
        
            <h3 className='auth__title'>ChiroNotes <i className=" far fa-edit"></i></h3>
        
        <form onSubmit={handleLogin} className={`animate__animated animate__fadeIn animate__faster`}>
            {    
              msgError &&                                           //realiza la operacion logica para la muestra o no de los errores
             
             <div className='auth__alert-error'>
             <p>{`${msgError}`}</p>
             </div>
                 
             }
        
        <input
        className='auth__input'
        type='text'
        placeholder='Email'
        name='email'
        autoComplete='off'
        value={email}
        onChange={handleInputChange}
        />
        <input
        className='auth__input'
        type={show ? 'text' : 'password'}
        placeholder='Contraseña'
        name='password'
        value={password}
        onChange={handleInputChange}
        />
        <div className='auth__checkbox'>
        <p className='auth__show-passwd'>Mostrar contraseña</p>
        <input type='checkbox'
         onChange={showOrHiddenPassword}/>
        </div>

        <button
        type='submit'
        className='btn btn-primary btn-block'
      
        >Ingresar <i className=' icons fas fa-arrow-right'></i>
        </button>
        
        <div className='auth__social-networks'>
            <p><b></b></p>
             
        <div className="google-btn" 
        onClick={handleGoogleLogin}
        >
            <div className="google-icon-wrapper">
                <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
                 <p className="btn-text">
                <b>Ingresa con google</b>
                     </p>
        </div>
        </div>
        <Link to='/auth/register'
        className='link'>
           Crear una cuenta nueva <i className="icons fa fa-angle-double-right"></i>
        </Link>

        </form>

             
    </div>
    
   
    )
}
