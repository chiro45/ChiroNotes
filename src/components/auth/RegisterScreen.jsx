import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator  from 'validator';
import {startRegisterWhithEmailPasswordName} from '../../actions/auth';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';




export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const {msgError} = useSelector(state=>state.ui)

    const [formValues, handleInputChange] = useForm({           //utilizamos el coustomHook para el manejo del formulario 
        name: 'fernando',                                               //declaramos el initialState
        email: 'fernando@gmail,cim',
        password: '123123',
        password2: '123123',
    });

    const {name, email , password, password2} = formValues;     //desestructuracion de los datos
    const handleRegister = (e)=>{
        e.preventDefault();
        if(isFormValid()){
            dispatch(startRegisterWhithEmailPasswordName(email, password, name)) //si no hay errores en el form se realiza la operacion

        }
    }

    
    const isFormValid = ()=>{                                                           //realiza la verificacion de los datos
        if(name.trim().length < 5 ){
            dispatch(setError('El nombre de usuario debe ser mayor a 5 caracteres'))
            return false;
        } else if(!validator.isEmail(email)){
            dispatch(setError('El email es incorrecto'))
            return false;
        }else if(password !== password2 || password.length < 8){
            dispatch(setError('La contrase単a es menor a 8 caracteres o estas no coinciden'))
            return false;
        }
        dispatch(removeError())
        return true;
    }
    return (
        <div>
            <h3 className='auth__title'>ChiroNotes <i className="  far fa-edit"></i></h3>
                
            
            <form
             onSubmit={handleRegister}
             className={`animate__animated animate__fadeIn animate__faster`}>

                 {    
                  msgError &&                   //realiza la comprobacion para mostrar en pantalla los errores o no
                 
                 <div className='auth__alert-error'>    
                 <p>{`${msgError}`}</p>
                 </div>
                     
                 }
            <input
            className='auth__input'
            type='text'
            placeholder='Nombre'
            name='name'
            autoComplete='off'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Email'
            name='email'
            autoComplete='off'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Contrase単a'
            name='password'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Confirma tu Contrase単a'
            name='password2'
            onChange={handleInputChange}
            />

            <button
            type='submit'
            className='btn btn-primary btn-block marginBottom-5'
           
            >Registrarse <i className=' icons fa fa-sign-in-alt'></i>
            </button>
            
            
            <Link to='/auth/login'
            className='link '>
              多Ya te has Registrado?
            </Link>

            </form>


        </div>
    )
}
