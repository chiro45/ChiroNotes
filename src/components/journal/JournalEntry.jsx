import React from 'react';
import moment from 'moment'
import PropTypes from 'prop-types';


import 'moment/locale/es'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({id,date,title,body,url}) => {
    const dispatch = useDispatch()
    const noteDate = moment(date);
    //click event sobre alguna de las notas
    const handleEntryClick = ()=>{

        //realizamos el disparo mandando el id y los demas datos de la nota
        dispatch(activeNote(id,{
            date,title,body,url
        })

        );
    }
   
    return (
      <div className='journal__entry pointer animate__animated animate__fadeInRight animate__fast'
      onClick={handleEntryClick}>
      { //en el caso de que la nota tenga imagen entonces se muestra el div
          url &&
      <div className='journal__entry-picture'
      style={{
          backgroundSize:'cover',
          backgroundImage:`url(${url})`,
      }}>
              
      </div>
    }
        
        <div className='journal__entry-body'>
        <p className='journal__entry-title'>
            {title}
        </p>
        <p className='journal__entry-content'>
            {body}
        </p>
        </div>

        <div className='journal__entry-date-box'>
            <span>{noteDate.format('dddd')}</span>
            <h4>{noteDate.format('Do')}</h4>
        </div>
      </div>
  )
};



JournalEntry.propTypes ={
    id: PropTypes.string,
    date: PropTypes.number,
    title: PropTypes.string,
    body: PropTypes.string,
    url: PropTypes.string
}