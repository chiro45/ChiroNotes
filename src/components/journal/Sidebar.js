
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';




export const Sidebar = () => {

  const dispatch = useDispatch();
  const {name} = useSelector(state => state.auth);
  
  const handleLogout=()=>{

    dispatch(startLogout())
  }


  const handleAddNew = ()=>{
    dispatch(startNewNote())
    
  }
  
  return (
  <aside className='journal__sidebar '>

      <div className='journal__sidebar-navbar '>
          <h3 className='marginTop-5'>
          
          <span className='journal__name'><i className="fas fa-user"></i> {name}</span>
          </h3>
      </div>

      <div className='journal__new-entry'
      onClick={handleAddNew}>
            <i className='far fa-calendar-plus fa-5x'></i>
            <p className='marginTop-5'> Nueva Entrada</p>
      </div>
      <JournalEntries />
        <button 
        className='btn-navBar'
        onClick={handleLogout}
        >
            Cerrar SesiÃ³n <i className="fas fa-sign-out-alt"></i>
        </button>
  </aside>
  )
};
