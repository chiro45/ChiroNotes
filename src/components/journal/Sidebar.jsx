
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';




export const Sidebar = () => {
 // declaramos el dispatch del redux
  const dispatch = useDispatch();
  //extraemos el name del stores del state.auth
  const {name , photoURL} = useSelector(state => state.auth);


  const state = useSelector(state => state);
 
  
 
  
  const handleLogout=()=>{

    dispatch(startLogout())
  }

  //activa el disparo del tartnNewNote
  const handleAddNew = ()=>{
    dispatch(startNewNote())
    
  }
  
  return (
  <aside className='journal__sidebar ' id='sidebar'>

      <div className='journal__sidebar-navbar '>
          <div className='journal__containerName'>
          
          {
             (photoURL)
              ? <div >  <img className='journal__img' src={photoURL}/> <p className='journal__name'>{name}</p></div>
              : <div >  <i className=" journal__img fas fa-user"></i> <p className='journal__name'>{name}</p></div>
              // : <span className='journal__name'>   {name}</span>
          }
          </div>
      </div>

      <div className='journal__new-entry'
      onClick={handleAddNew}>
            <i className='far fa-calendar-plus fa-5x'></i>
            <p className='marginTop-5'> Nueva Entrada</p>
      </div>
      <JournalEntries />
        <button 
        className='btn-navBar'
        onClick={handleLogout}
        >
            Cerrar Sesi√≥n <i className="fas fa-sign-out-alt"></i>
        </button>
  </aside>
  )
};
