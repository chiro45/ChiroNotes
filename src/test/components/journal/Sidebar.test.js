import { Sidebar } from "../../../components/journal/Sidebar";

import {mount} from 'enzyme'
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import thunk from "redux-thunk";
import '@testing-library/jest-dom'
import configureStore from 'redux-mock-store';

import { startNewNote } from "../../../actions/notes";
import { startLogout } from "../../../actions/auth";
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock("../../../actions/auth",()=>({
    startLogout : jest.fn()
}))
jest.mock('../../../actions/notes',()=>({
    startNewNote: jest.fn()
}))


const initState = {
    auth: {
        uid:'123123',
        name:'luciano',
        photoURL: null
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        active: null,
        notes:[]
        
    }
};

let store = mockStore(initState);
store.dispatch= jest.fn();


describe('Pruebas en el <Sidebar/>', () => {
    const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                      <Sidebar/>
                </MemoryRouter>
            </Provider>
    )
    
    test('debe de mostrarse correctamente', () => {
      expect(wrapper).toMatchSnapshot()
    });
    test('debe de llamar el startLogout', async() => {
        
            wrapper.find('button').prop('onClick')()

            expect(startLogout).toHaveBeenCalled()
      });
      test('debe de llamar el startNewNote', () => {
            wrapper.find('.journal__new-entry').prop('onClick')()

            expect(startNewNote).toHaveBeenCalled()
      });
        
});
