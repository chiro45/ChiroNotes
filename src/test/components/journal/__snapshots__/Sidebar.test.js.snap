// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <Sidebar/> debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar "
  id="sidebar"
>
  <div
    className="journal__sidebar-navbar "
  >
    <div
      className="journal__containerName"
    >
      <div>
          
        <i
          className=" journal__img fas fa-user"
        />
         
        <p
          className="journal__name"
        >
          luciano
        </p>
      </div>
    </div>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="marginTop-5"
    >
       Nueva Entrada
    </p>
  </div>
  <div
    className="journal__entries "
  />
  <button
    className="btn-navBar"
    onClick={[Function]}
  >
    Cerrar Sesi√≥n 
    <i
      className="fas fa-sign-out-alt"
    />
  </button>
</aside>
`;
